README.md
# 常用操作
### 查看账户充值表
* cleos get table card card balance

### 查看游戏表
* cleos get table card card games

### 查看action
* cleos get actions card -1 -j --console

### 查看transaction
* cleos get transaction 17f03952abcbd312ae54de8e5bb8680a589dac123ab44aee0e5e1f0128d2a372 

### 充值
* cleos push action token transfer '["alice", "card", "30.0000 EOS" "deposit-"]' -p alice@active

### 提币
* cleos push action token transfer '["alice", "card", "0.0001 EOS" "withdraw-"]' -p alice@active

### 开启一局游戏 
* cleos push action token transfer '[ "dev", "card", "0.0001 EOS" "start-2-5000-3-dev-bob-alice-" ]' -p dev@active

* 2 -- house id
  
* dev -- 游戏账号
 
* 玩家 dev bob alice, 实际dev是不参与游戏的，不想再创建账号, 这里直接使用了dev当做一个玩家

### 加注
* cleos push action token transfer '[ "alice", "card", "0.5000 EOS" "raise-2-" ]' -p alice@active

### 弃牌
* cleos push action token transfer '[ "alice", "card", "0.0001 EOS" "bust-2-" ]' -p alice@active

### 开牌
* cleos push action token transfer '[ "bob", "card", "0.0005 EOS" "open-2-dev-" ]' -p bob@active

### 删除games table 指定id的 record
* cleos push action card clean '{"id": 2}' -p card@active


# 计算牌型
使用开奖记录的种子“seed”进行 SHA512 运算，得到 A = sha512(seed)（64位长度的uint16数组），使用该数组即可完成发牌。

牌组的排序及相关值如下所示（52位长度的一维数组）：

* 0x02,0x03,0x04,0x05,0x06,0x07,0x08,0x09,0x0A,0x0B,0x0C,0x0D,0x0E //方块 2 - K - A
* 0x12,0x13,0x14,0x15,0x16,0x17,0x18,0x19,0x1A,0x1B,0x1C,0x1D,0x1E //梅花 2 - K - A
* 0x22,0x23,0x24,0x25,0x26,0x27,0x28,0x29,0x2A,0x2B,0x2C,0x2D,0x2E //红桃 2 - K - A
* 0x32,0x33,0x34,0x35,0x36,0x37,0x38,0x39,0x3A,0x3B,0x3C,0x3D,0x3E //黑桃 2 - K - A

发牌逻辑，最多总共需要发十五张牌，取牌按照如下公式计算得出，算出一张牌后，将该牌从牌组中移出以便计算下一张牌的位置：
假设有n个玩家：
* 前n张牌: A[i]*17+A[i+1]*13+A[i+2]*11; 
* 后2n张牌: A[i]*17+A[i+1]*13+A[i+2]

如果n=5, 五个玩家分别为: A B C D E, 15张牌计算如下：

(A[0]*17+A[1]*13+A[2]*11)%剩余牌的张数(52)      作为玩家A第一张牌位置，从牌组中取出。

(A[4]*17+A[5]*13+A[6]*11)%剩余牌的张数(51)      作为玩家B第一张牌位置，从牌组中取出。

(A[8]*17+A[9]*13+A[10]*11)%剩余牌的张数(50)     作为玩家C第一张牌位置，从牌组中取出。

(A[12]*17+A[13]*13+A[14]*11)%剩余牌的张数(49)   作为玩家D第一张牌位置，从牌组中取出。

(A[16]*17+A[17]*13+A[18]*11)%剩余牌的张数(48)   作为玩家E第一张牌位置，从牌组中取出。 

(A[20]*17+A[21]*13+A[22])%剩余牌的张数(47)      作为玩家A第二张牌位置，从牌组中取出。

(A[24]*17+A[25]*13+A[26])%剩余牌的张数(46)      作为玩家B第二张牌位置，从牌组中取出。

(A[28]*17+A[29]*13+A[30])%剩余牌的张数(45)      作为玩家C第二张牌位置，从牌组中取出。

(A[32]*17+A[33]*13+A[34])%剩余牌的张数(44)      作为玩家D第二张牌位置，从牌组中取出。

(A[36]*17+A[37]*13+A[38])%剩余牌的张数(43)      作为玩家E第二张牌位置，从牌组中取出。 

(A[40]*17+A[41]*13+A[42])%剩余牌的张数(42)      作为玩家A第三张牌位置，从牌组中取出。

(A[44]*17+A[45]*13+A[46])%剩余牌的张数(41)      作为玩家B第三张牌位置，从牌组中取出。

(A[48]*17+A[49]*13+A[50])%剩余牌的张数(40)      作为玩家C第三张牌位置，从牌组中取出。

(A[52]*17+A[53]*13+A[54])%剩余牌的张数(39)      作为玩家D第三张牌位置，从牌组中取出。

(A[56]17+A[57]*13+A[58])%剩余牌的张数(38)      作为玩家E第三张牌位置，从牌组中取出。






history
  552  cleos get currency balance token card EOS
  555  cleos push action token transfer '[ "bob", "card", "0.1200 EOS", "transfer test" ]' -p bob@active
  557  cleos push action token transfer '[ "bob", "card", "0.1200 EOS", "order-bob-1-1200-sell-" ]' -p bob@active
  558  cleos push action token transfer '[ "bob", "card", "0.1200 SYS", "order-bob-1-1200-sell-" ]' -p bob@active
  643  cleos push action token transfer '[ "bob", "card", "12.0000 EOS", "deposit-12EOS" ]' -p bob@active
  644  cleos push action token transfer '[ "bob", "card", "12.0000 SYS", "deposit-12EOS" ]' -p bob@active
  645  cleos push action token transfer '[ "bob", "card", "12.0000 EOS", "deposit-12EOS" ]' -p bob@active
  647  cleos push action token transfer '[ "bob", "card", "12.0000 EOS", "deposit-12EOS" ]' -p bob@active
  648  cleos push action token transfer '[ "bob", "card", "12.0000 EOS", "deposit" ]' -p bob@active
  649  cleos push action token transfer '[ "bob", "card", "12.0000 EOS", "deposit-12EOS" ]' -p bob@active
  650  cleos push action token transfer '[ "bob", "card", "12.0000 EOS", "deposit-12-" ]' -p bob@active
  651  cleos push action token transfer '[ "bob", "card", "12.0000 EOS", "deposit-" ]' -p bob@active
  652  cleos push action token transfer '[ "bob", "card", "12.0000 EOS", "deposit_" ]' -p bob@active
  653  cleos push action token transfer '[ "bob", "card", "12.0000 EOS", "deposit" ]' -p bob@active
  657  cleos push action token transfer '[ "bob", "card", "12.0000 EOS", "deposit" ]' -p bob@active
  658  cleos push action token transfer '[ "bob", "card", "12.0000 EOS", "deposit-12 eos" ]' -p bob@active
  661  cleos push action token transfer '[ "bob", "card", "12.0000 EOS", "deposit-12 eos" ]' -p bob@active
  664  cleos push action token transfer '[ "bob", "card", "12.0000 EOS", "deposit-12 eos" ]' -p bob@active
  667  cleos push action token transfer '[ "bob", "card", "12.0000 EOS", "deposit-12 eos" ]' -p bob@active
  670  cleos push action token transfer '[ "bob", "card", "12.0000 EOS", "deposit-12 eos" ]' -p bob@active
  673  cleos push action token transfer '[ "bob", "card", "12.0000 EOS", "deposit-12 eos" ]' -p bob@active
  676  cleos push action token transfer '[ "bob", "card", "10.0000 EOS", "deposit-10 eos" ]' -p bob@active
  679  cleos push action token transfer '[ "bob", "card", "10.0000 EOS", "deposit-10 eos" ]' -p bob@active
  683  cleos push action token transfer '[ "bob", "card", "10.0000 EOS", "deposit-10 eos" ]' -p bob@active
  685  cleos push action token transfer '[ "bob", "card", "0.0001 EOS", "withdraw-" ]' -p bob@active
  686  cleos push action token transfer '[ "bob", "card", "0.0001 EOS", "withdraw-" ]' -p bob@active
  687  cleos push action token transfer '[ "alice", "card", "10.0000 EOS", "deposit-10 eos" ]' -p alice@active
  691  cleos push action token transfer '[ "alice", "card", "10.0000 EOS", "deposit-10 eos" ]' -p alice@active
  693  cleos push action token transfer '[ "alice", "card", "10.0000 EOS", "deposit-10 eos" ]' -p alice1@active
  702  cleos push action token transfer '[ "alice", "card", "10.0000 EOS", "deposit-10 eos" ]' -p alice1@active
  703  cleos push action token transfer '[ "alice", "card", "10.0000 EOS", "deposit-10 eos" ]' -p alice@active
  705  cleos push action token transfer '[ "alice", "card", "0.0001 EOS" "withdraw-" ]' -p alice@active
  706  cleos push action token transfer '[ "alice", "card", "0.0001 EOS" "withdraw-" ]' -p alice@active
  709  cleos push action token transfer '[ "alice", "card", "0.0001 EOS" "withdraw-" ]' -p alice@active
  713  cleos push action token transfer '[ "alice", "card", "0.0001 EOS" "withdraw-" ]' -p alice@active
  714  cleos push action token transfer '[ "alice", "card", "10.0000 EOS", "deposit-10 eos" ]' -p alice@active
  715  cleos push action token transfer '[ "alice", "card", "12.0000 EOS", "deposit-12 eos" ]' -p alice@active
  716  cleos push action token transfer '[ "bob", "card", "12.0000 EOS", "deposit-12 eos" ]' -p bob@active
  718  cleos push action token transfer '[ "alice", "card", "0.0001 EOS" "withdraw-" ]' -p alice@active
  719  cleos push action token transfer '[ "dev", "card", "0.0001 EOS" "withdraw-" ]' -p dev@active
  720  cleos push action token transfer '[ "dev", "card", "12.0000 EOS" "deposit-" ]' -p dev@active
  721* cleos push action token transfer '[ "debug", "card", "12.0000 EOS" "deposit-" ]' -p @active
  722  cleos push action token transfer '[ "debug", "card", "12.0000 EOS" "deposit-" ]' -p debug@active
  731  cleos push action token transfer '[ "dev", "card", "0.0001 EOS" "start-0-5000-3-dev-bob-alice-" ]' -p dev@active
  732  cleos push action token transfer '[ "dev", "card", "8.0000 EOS" "deposit-" ]' -p dev@active
  734  cleos push action token transfer '[ "alice", "card", "14.0000 EOS" "deposit-" ]' -p alice@active
  739  cleos push action token transfer '[ "dev", "card", "0.0001 EOS" "start-0-5000-3-dev-bob-alice-" ]' -p dev@active
  740  cleos push action token transfer '[ "bob", "card", "0.0001 EOS" "start-0-5000-3-dev-bob-alice-" ]' -p bob@active
  744  cleos push action token transfer '[ "bob", "card", "0.0001 EOS" "start-0-5000-3-dev-bob-alice-" ]' -p bob@active
  745  cleos push action token transfer '[ "dev", "card", "0.0001 EOS" "start-0-5000-3-dev-bob-alice-" ]' -p dev@active
  751  cleos push action token transfer '[ "dev", "card", "0.0001 EOS" "start-0-5000-3-dev-bob-alice-" ]' -p dev@active
  754  cleos push action token transfer '[ "dev", "card", "0.0001 EOS" "start-0-5000-3-dev-bob-alice-" ]' -p dev@active
  757  cleos push action token transfer '[ "dev", "card", "0.0001 EOS" "start-0-5000-3-dev-bob-alice-" ]' -p dev@active
  760  cleos push action token transfer '[ "dev", "card", "0.0001 EOS" "start-0-5000-3-dev-bob-alice-" ]' -p dev@active
  763  cleos push action token transfer '[ "dev", "card", "0.0001 EOS" "start-0-5000-3-dev-bob-alice-" ]' -p dev@active

test.cpp
// #pragma once

#include "picosha2.h"

#include <iostream>
#include<sstream>
#include <string>
#include <vector>
#include <map>



using namespace std;

// //发牌数组
// int cards52[52] = {
//         0x02,0x03,0x04,0x05,0x06,0x07,0x08,0x09,0x0A,0x0B,0x0C,0x0D,0x0E, //方块 2 - K - A
//         0x12,0x13,0x14,0x15,0x16,0x17,0x18,0x19,0x1A,0x1B,0x1C,0x1D,0x1E, //梅花 2 - K - A
//         0x22,0x23,0x24,0x25,0x26,0x27,0x28,0x29,0x2A,0x2B,0x2C,0x2D,0x2E, //红桃 2 - K - A
//         0x32,0x33,0x34,0x35,0x36,0x37,0x38,0x39,0x3A,0x3B,0x3C,0x3D,0x3E, //黑桃 2 - K - A
// };

// enum flowertype {
//     DIAMOND = 0,            // 方片
//     CLUB,                   // 梅花
//     HEART,                  // 红桃
//     SPADE                   // 黑桃
// };

// enum cardnum { 
//     card2 = 2, card3 = 3, card4 = 4, card5 = 5, card6 = 6, card7 = 7, card8 = 8, 
//     card9 = 9, card10 = 10, cardJ = 11, cardQ = 12, cardK = 13, cardA = 14
// };

// enum playertype {
//     NORMAL = 0,             //普通牌
//     DOUBLE,                 //对子
//     STRAIGHT,               //顺子
//     FLUSH,                  //同花
//     STRAIGHT_FLUSH,         //同花顺
//     BOMB                    //炸弹
// };

// enum playertypemax {
//     NORMAL_MAX_VALUE            = 3803,     //普通牌: 最大值AKJ=14*16*16+13*16+11=3803
//     DOUBLE_MAX_VALUE            = 4040,     //对子牌: 最大值AAK=14*16+13=237,加上普通牌的基数3803=4040
//     STRAIGHT_MAX_VALUE          = 4052,     //顺子牌: 最大值AKQ=12，加上对子的最大值基数4040=4052
//     FLUSH_MAX_VALUE             = 7855,     //同花牌: 最大值AKJ，14*16*16+13*16+11=3803，加上顺子4052=7855
//     STRAIGHT_FLUSH_MAX_VALUE    = 7867,     //同花顺: 最大值AKQ=12，加上同花7855=7867
//     BOMB_MAX_VALUE              = 7881      //炸弹牌: 最大值AAA=14，加上同花顺7867=7881
// };

// struct card {
//     uint8_t                     flower;     //花色: flowertype
//     uint8_t                     number;     //牌点数: cardnum
// };

// struct playercard {
//     vector<card>                cards;      
//     uint8_t                     cardtype;  //牌类型：playtype
//     bool                        a32;       //是否是A32
//     int                         value;     //牌绝对值大小
// };

// vector<int> parsesha256(const string& seeds) {
//     vector<int> seed;
//     for(int i = 0; i < seeds.size(); i++) {
//         int n = seeds[i] - '0';
//         if(n >= 48) {
//             n = seeds[i] -'a' + 10;
//         }
//         seed.push_back(n);
//     }
//     return seed;
// }

// map<string, playercard> getRandomCards(const vector<string>& players, const string& seeds) {
//     vector<int> cards(cards52, cards52+52);
//     vector<int> seed = parsesha256(seeds);
//     int size = players.size();
//     int cardsize = cards.size();
//     map<string, playercard> playercards;
//     for(size_t i = 0; i < size*3; i++) {
//         int d = (i < size) ? 11 : 1;
//         int index = (seed[i]*17+seed[i+1]*13+seed[i+2]*d)%cardsize;
//         int random = cards[index];
//         card c;
//         c.flower = random/16;
//         c.number = random%16;
//         if(i < size) {
//             playercard pc;
//             pc.cards.push_back(c);
//             playercards.insert(std::pair<string, playercard>(players[i], pc));
//         } else {
//             string player = players[i%size];
//             playercards[player].cards.push_back(c);
//         }
//         std::remove(cards.begin(), cards.end(), random);
//         cardsize--;
//     }
//     return playercards;
// }

// string seedHash() {
//     string text = "this is a test. random hash text. ff3f4036a1164d1ddbad5b3edf9022addb3e1961a54a922708a6c1ffc49e5489 ";
//     time_t t = time(NULL);  //秒时间
//     char buf[128] = {0};
//     tm* gmt = gmtime(&t);  //转为格林威治时间  
//     strftime(buf, 64, "%Y-%m-%d %H:%M:%S", gmt);  
//     text += string(buf);
//     return picosha2::hash256_hex_string(text);
// }

int main()
{
    istringstream istream;

    // //string seed = "96f06aec039a44b9b053079476554fd6954ecd5a126692463cc3dab112b172199a3ea8a32ab998156ec47157b1bb595d3aa88277a89cb2a78302668d8560d50e";
    // string seed = "ff3f4036a1164d1ddbad5b3edf9022addb3e1961a54a922708a6c1ffc49e5489";
    // for(int i = 0; i < seed.size(); i++) {
    //     // istream.str(std::to_string(seed[i]));
    //     // int n;
    //     // istream >> n;
    //     // cout << "n = " << n << endl;

    //     int n = seed[i] - '0';
    //     if(n >= 48) {
    //         n = seed[i] -'a' + 10;
    //     }
    //     cout << n << " ";
    // }

    // cout << "seed size: " << seed.size() << endl;

    //string seed = "ff3f4036a1164d1ddbad5b3edf9022addb3e1961a54a922708a6c1ffc49e5489";
    string src_str = "this is a test. random hash text. ff3f4036a1164d1ddbad5b3edf9022addb3e1961a54a922708a6c1ffc49e5489 ";
    time_t t = time(NULL);  //秒时间
    char buf[128]= {0};
    tm* gmt = gmtime(&t);  //转为格林威治时间  
    strftime(buf, 64, "%Y-%m-%d %H:%M:%S", gmt);  
    std::cout << buf << std::endl; 
    src_str += string(buf);
    cout << "src_str: " << src_str << endl;
    string hash_hex_str = picosha2::hash256_hex_string(src_str);
    cout << "hash_hex_str: " << hash_hex_str << endl;
    string seed = hash_hex_str;
    for(int i = 0; i < seed.size(); i++) {
        int n = seed[i] - '0';
        if(n >= 48) {
            n = seed[i] -'a' + 10;
        }
        cout << n << " ";
    }
    cout << "seed size: " << seed.size() << endl;

    // vector<string> players;
    // players.push_back("Jane");
    // players.push_back("Lisa");
    // players.push_back("Peter");
    // map<string, playercard> cards = getRandomCards(players, seed);
    // for(auto iter = cards.begin(); iter != cards.end(); iter++) {
    //     cout << iter->first << ", cards:  " << endl;
    //     auto player = iter->second.cards;
    //     for(auto card = player.begin(); card != player.end(); card++) {
    //         cout << "flower: " << card->flower << " | card: " << card->number << endl;
    //     }
    //     cout << endl << endl;
    // }

    return 0;
}

threecard.hpp
#pragma once

#include "utils.hpp"

#include <eosio/eosio.hpp>
#include <eosio/asset.hpp>
#include <eosio/time.hpp>
#include <eosio/system.hpp>
#include <eosio/singleton.hpp>

#include <string>

using namespace eosio;
using namespace std;

CONTRACT threecard : public eosio::contract {
    public:
        using contract::contract;

        threecard(name receiver, name code, datastream<const char*> ds)
            : contract(receiver, code, ds),
              _global(receiver, receiver.value),
              _games(receiver, receiver.value),
              //_balance(receiver, receiver.value),
              _gamedata(receiver, receiver.value),
              _playerdata(receiver, receiver.value)
        {}

        //table: global, game, deposit
        TABLE global {
            uint64_t                    current_id;
        };

        TABLE game {
            uint64_t                    id;             //house id
            uint8_t                     size;           //surplus player number
            map<string, bool>           players;        //player and player status
            map<string, playercard>     cards;          //public cards
            asset                       amount;         //fund pool
            string                      winner;         //current winner
            time_point_sec              created_at;     //house create time

            uint64_t primary_key() const { return id; }
        };

        // TABLE balance {
        //     name                        player;
        //     asset                       balance;

        //     uint64_t primary_key() const { return player.value; }
        // };

        TABLE playerdata {
            name                        player;
            asset                       balance;
            uint64_t                    count;

            uint64_t primary_key() const { return player.value; }
        };

        //private table
        struct gamedata {
            uint64_t                    id;             //game id
            string                      seed;           //random cards use
            map<string, playercard>     cards;          //surplus cards

            uint64_t primary_key() const { return id; }
        };

        typedef singleton<N(global), global>         tb_global;
        typedef multi_index<N(games), game>          tb_games;
        //typedef multi_index<N(balance), balance>     tb_balance;
        typedef multi_index<N(gamedata), gamedata>   tb_gamedata;
        typedef multi_index<N(players), playerdata>  tb_playerdata;
        //table end

    private:
        tb_global                   _global;
        tb_games                    _games;
        //tb_balance                  _balance;
        tb_gamedata                 _gamedata;
        tb_playerdata               _playerdata;

    private:
        void parse_memo(string memo, string& action, map<string, string>& params);
        void parse_memo_start(string memo, uint64_t& id, uint64_t& volume, vector<string>& players);
        void parse_memo_raise_or_bust(string memo, uint64_t& id);
        void parse_memo_open_card(string memo, uint64_t& id, string& player);

        // uint64_t next_id() {
        //     global g = _global.get_or_default(
        //         global{.current_id = _games.available_primary_key()});
        //     g.current_id += 1;
        //     _global.set(g, _self);
        //     return g.current_id;
        // }

        //充值
        void deposit(const name& player, uint64_t quantity, bool flag = false);

        //提币
        void withdraw(const name& player);

        //game start
        void start(uint64_t id, const vector<string>& players, uint64_t quantity, const string& seed);

        //加注
        void raise(uint64_t id, const name& player, const asset& quantity);

        //弃牌
        void bust(uint64_t id, const string& player);

        //比牌
        void open(uint64_t id, const string& from, const string& to, uint64_t quantity);

public:

        ACTION receipt(const game& data) {
            //require(_self, "game over receipt error");
        }

        ACTION depositrp(const playerdata& data, const string& memo) {
            //require(_self, "deposit receipt error");
            //require_auth(_self);
        }  //deposit receipt

        //clean games, balance table //for test
        ACTION clean(const uint64_t& id) {
            auto existing = _games.find(id);
            if(existing != _games.end()) {
                _games.erase(existing);
            }
        }

        void transfer(const name& from, const name& to, const asset& quantity, 
            const string& memo);
};



utils.hpp
#include <eosio/eosio.hpp>
#include <eosio/asset.hpp>
#include <eosio/time.hpp>
#include <eosio/system.hpp>
#include <eosio/singleton.hpp>

#include <string>
#include <map>
#include <vector>
#include <list>
#include <algorithm>
#include <utility>

using namespace eosio;
using namespace std;


#define N(X)            eosio::name(#X)
#define S(P,X)          eosio::symbol(#X,P)
#define EOS_SYMBOL      S(4, EOS)

#define TRANSFER_CODE   token       //eosio.token
#define FEE             1           //0.0001 eos
#define GAME_ACCOUNT    name("dev")

//发牌数组
int cards52[52] = {
        0x02,0x03,0x04,0x05,0x06,0x07,0x08,0x09,0x0A,0x0B,0x0C,0x0D,0x0E, //方块 2 - K - A
        0x12,0x13,0x14,0x15,0x16,0x17,0x18,0x19,0x1A,0x1B,0x1C,0x1D,0x1E, //梅花 2 - K - A
        0x22,0x23,0x24,0x25,0x26,0x27,0x28,0x29,0x2A,0x2B,0x2C,0x2D,0x2E, //红桃 2 - K - A
        0x32,0x33,0x34,0x35,0x36,0x37,0x38,0x39,0x3A,0x3B,0x3C,0x3D,0x3E, //黑桃 2 - K - A
};

enum flowertype {
    DIAMOND = 0,            // 方片 ♦
    CLUB,                   // 梅花 ♣
    HEART,                  // 红桃 ♥
    SPADE                   // 黑桃 ♠
};

enum cardnum { 
    card2 = 2, card3 = 3, card4 = 4, card5 = 5, card6 = 6, card7 = 7, card8 = 8, 
    card9 = 9, card10 = 10, cardJ = 11, cardQ = 12, cardK = 13, cardA = 14
};

enum playertype {
    NORMAL = 0,             //普通牌
    DOUBLE,                 //对子
    STRAIGHT,               //顺子
    FLUSH,                  //同花
    STRAIGHT_FLUSH,         //同花顺
    BOMB                    //炸弹
};

enum playertypemax {
    NORMAL_MAX_VALUE            = 3803,     //普通牌: 最大值AKJ=14*16*16+13*16+11=3803
    DOUBLE_MAX_VALUE            = 4040,     //对子牌: 最大值AAK=14*16+13=237,加上普通牌的基数3803=4040
    STRAIGHT_MAX_VALUE          = 4052,     //顺子牌: 最大值AKQ=12，加上对子的最大值基数4040=4052
    FLUSH_MAX_VALUE             = 7855,     //同花牌: 最大值AKJ，14*16*16+13*16+11=3803，加上顺子4052=7855
    STRAIGHT_FLUSH_MAX_VALUE    = 7867,     //同花顺: 最大值AKQ=12，加上同花7855=7867
    BOMB_MAX_VALUE              = 7881      //炸弹牌: 最大值AAA=14，加上同花顺7867=7881
};

struct card {
    uint8_t                     flower;     //花色: flowertype
    uint8_t                     number;     //牌点数: cardnum
};

struct playercard {
    vector<card>                cards;      
    uint8_t                     cardtype;  //牌类型：playtype
    bool                        a32;       //是否是A32
    int                         value;     //牌绝对值大小
};

size_t sub2sep( const std::string& input, std::string& output, const char& separator,
        size_t first_pos = 0, bool required = false ) {
    check(first_pos != std::string::npos, "invalid first pos");
    auto pos = input.find(separator, first_pos);
    if (pos == std::string::npos) {
        check(!required, "parse memo error");
        return std::string::npos;
    }
    output = input.substr(first_pos, pos - first_pos);
    return pos;
}

//获取玩家牌的绝对值大小
//cards已经按照牌值从大到小排序：888 | AKQ, 987, A32 | AKJ, 965 | AKQ, 987, A32 | AAJ 996
//a, 获取炸弹牌值绝对大小
int getBombValue(const playercard& player) {
    return player.cards[0].number + STRAIGHT_FLUSH_MAX_VALUE;
}

//b, 获取同花顺牌值绝对大小, 特殊：A32也是同花顺，是最小的同花顺
int getStraightFlushValue(const playercard& player) {
    if (player.a32) {
        return 1 + FLUSH_MAX_VALUE;  //此时A就等于是1
    }
    return player.cards[2].number + FLUSH_MAX_VALUE;
}

//c, 获取同花牌值绝对大小
int getFlushValue(const playercard& player) {
    return player.cards[0].number * 256 + player.cards[1].number * 16 + player.cards[2].number
            + STRAIGHT_MAX_VALUE;
}

//d, 获取顺子牌值绝对大小
int getStraightValue(const playercard& player) {
    if (player.a32) {
        return 1 + DOUBLE_MAX_VALUE;   //此时A就等于是1
    }
    return player.cards[2].number + DOUBLE_MAX_VALUE;
}

//e, 获取对子牌值绝对大小
// 在判断牌型时，如果是对子，则将对子放在数组前面两位
int getDoubleValue(const playercard&  player) {
    return player.cards[1].number * 16 + player.cards[2].number + NORMAL_MAX_VALUE;
}

//f, 获取普通牌值绝对大小
int getNormalValue(const playercard&  player) {
    return player.cards[0].number * 256 + player.cards[1].number * 16 + player.cards[2].number;
}

//玩家牌排序
//a, 将一副牌按牌面从大到小排序
void sortPlayerByNumber(playercard&  player) {
    for (size_t i = 0; i < 3; i++) {
        for (size_t j = 0; j < 3 - i - 1; j++) {
            if (player.cards[j].number < player.cards[j + 1].number) {
                card tempCard =player.cards[j];
                player.cards[j] = player.cards[j + 1];
                player.cards[j + 1] = tempCard;
            }
        }
    }
}

//b, 将对子移动到一幅牌的前两位
void moveDouble2Front(playercard&  player) {
    if (player.cards[1].number == player.cards[2].number) {
        card tempCard = player.cards[0];
        player.cards[0] = player.cards[2];
        player.cards[2] = tempCard;
    }
}

//牌类型判断, cards已经按照牌值从大到小排序：888 | AKQ, 987, A32 | AKJ, 965 | AKQ, 987, A32 | AAJ 996
// a, 是否炸弹
bool isBomb(const playercard& player) {
    return player.cards[0].number == player.cards[1].number
            && player.cards[1].number == player.cards[2].number;
}

// b, 是否同花
bool isFlush(const playercard& player) {
    return player.cards[0].flower == player.cards[1].flower
            && player.cards[1].flower == player.cards[2].flower;
}

// c, 是否是A23
bool isA32(const playercard& player) {
    return player.cards[0].number == 14 && player.cards[1].number == 3
        && player.cards[2].number == 2;
}

// d, 是否顺子, A32也是同花顺，是最小的同花顺
bool isStraight(const playercard& player) {
    bool isNormalStraight = player.cards[0].number == player.cards[1].number + 1
            && player.cards[1].number == player.cards[2].number + 1;
    return isNormalStraight || isA32(player);
}

// e, 是否对子
bool isDouble(const playercard& player) {
    return player.cards[0].number == player.cards[1].number
            || player.cards[1].number == player.cards[2].number;
}

//牌值计算：判断玩家牌型、计算牌型绝对值大小
void calcCardsValueAndType(playercard& player) {
    if (isFlush(player)) {
        if (isStraight(player)) {
            // 同花顺
            player.cardtype = STRAIGHT_FLUSH;
            player.value = getStraightFlushValue(player);
        } else {
            // 同花
            player.cardtype = FLUSH;
            player.value = getFlushValue(player);
        }
    } else if (isStraight(player)) {
        // 顺子
        player.cardtype = STRAIGHT;
        player.value = getStraightValue(player);
    } else if (isDouble(player)) {
        if (isBomb(player)) {
            // 炸弹
            player.cardtype = BOMB;
            player.value = getBombValue(player);
        } else {
            // 对子 
            moveDouble2Front(player);  //将对子放到玩家牌的前两张的位置,以便于之后的牌值计算
            player.cardtype = DOUBLE;
            player.value = getDoubleValue(player);
        }
    } else {
        // 普通牌
        player.cardtype = NORMAL;
        player.value = getNormalValue(player);
    }
}

//玩家牌排序
void sortPlayersCards(map<string, playercard>& playerCards) {
    auto iter = playerCards.begin();
    for(; iter != playerCards.end(); iter++) {
        auto player = iter->second;
        sortPlayerByNumber(player);
        calcCardsValueAndType(player);
        playerCards[iter->first] = player;
    }
}

//玩家比牌: player1 compare player2
bool cardCompare(const playercard& player1, const playercard& player2) {
    return player1.value > player2.value;
}

vector<int> parsesha256(const string& seeds) {
    vector<int> seed;
    for(int i = 0; i < seeds.size(); i++) {
        int n = seeds[i] - '0';
        if(n >= 48) {
            n = seeds[i] -'a' + 10;
        }
        seed.push_back(n);
    }
    return seed;
}

//前n张牌:A[i]*17+A[i+1]*13+A[i+2]*11; 后2n张牌: A[i]*17+A[i+1]*13+A[i+2]
//seed sha256 64bytes 64个16进制数
map<string, playercard> getRandomCards(const vector<string>& players, const string& seeds) {
    vector<int> cards(cards52, cards52+52);
    vector<int> seed = parsesha256(seeds);
    int size = players.size();
    int cardsize = cards.size();
    map<string, playercard> playercards;
    for(size_t i = 0; i < size*3; i++) {
        int d = (i < size) ? 11 : 1;
        int index = (seed[i]*17+seed[i+1]*13+seed[i+2]*d)%cardsize;
        int random = cards[index];
        card c;
        c.flower = random/16;
        c.number = random%16;
        if(i < size) {
            playercard pc;
            pc.cards.push_back(c);
            playercards.insert(std::pair<string, playercard>(players[i], pc));
        } else {
            string player = players[i%size];
            playercards[player].cards.push_back(c);
        }
        std::remove(cards.begin(), cards.end(), random);
        cardsize--;
    }
    sortPlayersCards(playercards);
    return playercards;
}


